### ISSUE TITLE
iOS: Cold boot should preserve feeds and icons

### ISSUE THREAD

**Issue by samuelclay** [samuelclay/NewsBlur#1797]
See this video, I get this basically every morning. It's supposed to load All Site Stories automatically but it's loading before it should. The app used to store feeds and icons indefinitely, so let's figure out how to get back to that system, even if the data is stale.

https://github.com/samuelclay/NewsBlur/assets/44229/7751024b-7f3b-4423-b1d0-4092e3a7fd3c

**Comment by Dejal** [comment]
Please try this and see if it helps (in the Dejal branch). I suspect a race condition between loading the the view and preparing the icon cache.

I'm not able to reproduce the issue on my iPhone or the Simulator, so I can't confirm a fix. If this doesn't solve it, I'll investigate further.

**Comment by samuelclay** [comment]
I'm releasing `dejal` as v12.1.0.

**Comment by samuelclay** [comment]
This is still happening with seemingly no difference. 

https://github.com/samuelclay/NewsBlur/assets/44229/eb018b84-d2dd-4310-8e12-2cd240fde414

Try logging in as me on the dev build and opening the app on a cold boot. This doesn't happen every time but if there's a 2 second pause, it's guaranteed. Try adding sleeps to repro it.

**Comment by samuelclay** [comment]
This is really a top priority, as it makes me not want to open the app as I know I'll have to wait for sites and icons to load instead of diving right in.

**Comment by Dejal** [comment]
I was scheduled to work on NewsBlur next week, but thought I'd tackle this now, since it's bothering you, and it is proving tricky to track down.

I've spent some time investigating the launch sequence and image caching. I've fixed another possibly related issue, where the views were being loaded before the cache was ready. See if that helps.

I tried adding sleeps to repro it, but couldn't find any location that triggered it.

I've also added a logging line to see if the cache is finding the images. Run the app from Xcode and search the log for "🌄". It will show some images as not cached, but should find some of them.

There's a commented out line (appropriately line 666 of PINDiskCache) that when uncommented will simulate this issue by not finding anything in the cache. I have examined the cache code, and can't see why it might fail; there is an async part of the initialization that raised a red flag, but the app will block until that is complete, so that seems fine. It really feels like a cache issue, though.

Anyway, try this, and see if it is any better, and look at that logging to see if that indicates an issue.

I will continue to investigate.

**Comment by samuelclay** [comment]
I recommend logging in as me and then using the app. Since I have 900 sites, it makes it clear.

Also, I noticed that the app will load "All Site Stories" in about a second, even if nothing has come in from the server. That shouldn't be happening.

**Comment by Dejal** [comment]
I did log in as you. I still couldn't reproduce it.

Have you tried it again with that change? Is the issue still occurring? Have you looked at the logging? (Just paste 🌄 into the filter field below the console log.)

As mentioned, the issue seems to be with loading from the disk cache; it then successfully fetches the icons from the server into the memory cache.

If you turn off loading a folder on launch, does that help?

I agree it shouldn't go to the folder until it has loaded the update; I'll look into that too.

**Comment by samuelclay** [comment]
It doesn't happen every time, only after my phone has charged overnight or enough time has passed. It's like there's two types of cold boots: 1) after force quitting the app, it boots quickly and pre-loads images and feeds, and 2) after waiting a few hours, it has lost the local cache of feeds and icons.

**Comment by Dejal** [comment]
Yeah, that makes it especially hard to track down (and hard for you to confirm a fix). When I try changing something and run the app, it's like after a force quit. The second situation is after being evicted from memory; kinda a hybrid of running and not running.

I'll keep trying to repro it, or failing that to spot other possible causes. I may not get anything else done in this work block, but I assume this issue is annoying enough to be worth it to you.

**Comment by Dejal** [comment]
I spent some more time today investigating further, without further progress. Frustrating.

I could (sometimes) force an eviction by refusing the complete a background network request, but it still found the icons in the cache.

I did change the "when opening app" behavior to do so once the data has been received, instead of just offline data. (I think that might have been an intentional choice when originally implemented, but waiting for new stories makes more sense.)

So I'm going to put this on hold for a few days, to see if I can come up with any other ideas. Please try running the app, and see if the issue still occurs for you.

**Comment by samuelclay** [comment]
Yeah I think we should change that back and be smarter about what constitutes offline data. I want it to immediately open the selected folder IF there are offline stories. Let's just check if there actually are offline stories.

**Comment by Dejal** [comment]
Try that on for size, and let me know if the folder opening behavior works for you, and if the icon issue still occurs (see the details in the commit message). I'll come back to this next week if necessary.

_(Needless to say, you shouldn't release this publicly until the hack is removed. Also, Xcode Cloud is suspended until you review Apple's latest agreement.)_

**Comment by samuelclay** [comment]
Looks like the icon cache was empty this morning

![image](https://github.com/samuelclay/NewsBlur/assets/44229/6ff6e162-a1e5-45c7-9b02-7ba4382915ca)

**Comment by Dejal** [comment]
Hmm. Maybe I should try using a different directory for the cache, if iOS is clearing the cache folder.

Is your phone low on space or something? I don't know why else iOS might be doing that, other than trying to free up disposable disk space.

**Comment by samuelclay** [comment]
Can we store it in a DB instead of a cache directory? Same with the feed list and user avatar.

**Comment by samuelclay** [comment]
And I have 4GB free of 128GB

**Comment by Dejal** [comment]
I think big data like images is better in the file system. But there are other directories that iOS won't stomp on.

4GB free is pretty much full, so could well be the cause. But don't try to make more room yet; we should get this working even in a tight space situation.

**Comment by Dejal** [comment]
Alrighty, the app will now store its data in the Documents folder, instead of Caches, to avoid losing the data if iOS feels like freeing up some space. It's too important to the experience to leave to the whims of the OS; you're not the only person running with a full phone. 

The existing data is copied to the new location (rather than moved, so downgrading will still work; iOS will clear the cache later).

I've removed the logging and such, and am closing this issue, since I'm confident that this will fix the issue.

**Comment by samuelclay** [comment]
Yep, looks like this is now fixed!
